discovery-handlers/udev/src/discovery_handler.rs

```diff
@@ -68,37 +71,64 @@ impl DiscoveryHandler for DiscoveryHandlerImpl {
+                let mut devpaths: HashMap<String, Vec<DeviceProperties>> = HashMap::new();
                 udev_rules.iter().for_each(|rule| {
                     let enumerator = udev_enumerator::create_enumerator();
                     let paths = do_parse_and_find(enumerator, rule).unwrap();
-                    paths.into_iter().for_each(|path| {
-                        devpaths.insert(path);
-                    });
+                    for path in paths.into_iter() {
+                        if !discovery_handler_config.group_recursive {
+                            devpaths.insert(path.0.clone(), vec![path]);
+                        } else {
+                            match get_device_relatives(&path.0, devpaths.keys()) {
+                                (Some(parent), _) => devpaths.get_mut(&parent).unwrap().push(path),
+                                (None, childrens) => {
+                                    let id = path.0.clone();
+                                    let mut children_devices: Vec<DeviceProperties> = childrens
+                                        .into_iter()
+                                        .flat_map(|child| {
+                                            devpaths.remove(&child).unwrap().into_iter()
+                                        })
+                                        .collect();
+                                    children_devices.push(path);
+                                    let _ = devpaths.insert(id, children_devices);
+                                }
+                            }
+                        }
+                    }
                 });
                 trace!(
                     "discover - mapping and returning devices at devpaths {:?}",
                     devpaths
                 );
                 let discovered_devices = devpaths
                     .into_iter()
-                    .map(|(path, node)| {
-                        let mut properties = std::collections::HashMap::new();
+                    .map(|(id, paths)| {
+                        let mut properties = HashMap::new();
                         let mut mounts = Vec::new();
-                        if let Some(devnode) = node {
-                            properties
-                                .insert(super::UDEV_DEVNODE_LABEL_ID.to_string(), devnode.clone());
-                            mounts.push(Mount {
-                                container_path: devnode.clone(),
-                                host_path: devnode,
-                                read_only: true,
-                            });
+                        for (i, (_, node)) in paths.into_iter().enumerate() {
+                            let property_suffix = match discovery_handler_config.group_recursive {
+                                true => format!("_{}", i),
+                                false => "".to_string(),
+                            };
```

Comment:
Believe this can be simplified like so
```suggestion
            let property_suffix = discovery_handler_config.group_recursive.then(|| format!("_{}", i)).unwrap_or_default();
```